### PR TITLE
validate whether the sharding rules of the binding tables are consistent

### PR DESCRIPTION
**Pull request by cheese8**
Ref  #13596.

Changes proposed in this pull request:
- checkTheSameActualDataSourceNames per  bindingTableGroup while createBindingTableRule
- checkTheSameDatabaseShardingStrategy per  bindingTableGroup while createBindingTableRule
- checkTheSameTableShardingStrategy per  bindingTableGroup while createBindingTableRule

### CODE REVIEW

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 1]
```diff
@@ -192,11 +194,65 @@ private String getDefaultGenerateKeyColumn(final KeyGenerateStrategyConfiguratio
     private BindingTableRule createBindingTableRule(final String bindingTableGroup) {
         Map<String, TableRule> tableRules = Splitter.on(",").trimResults().splitToList(bindingTableGroup).stream()
                 .map(this::getTableRule).collect(Collectors.toMap(each -> each.getLogicTable().toLowerCase(), Function.identity(), (oldValue, currentValue) -> oldValue, LinkedHashMap::new));
+        checkTheSameActualDataSourceNames(tableRules, bindingTableGroup);
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:197]: @cheese8 Can you merge this three check method to checkBindingTableRule(tableRules)?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 2]
```diff
@@ -192,11 +194,65 @@ private String getDefaultGenerateKeyColumn(final KeyGenerateStrategyConfiguratio
...
+        checkTheSameShardingStrategyConfiguration(tableRules, bindingTableGroup, true);
+        checkTheSameShardingStrategyConfiguration(tableRules, bindingTableGroup, false);
         BindingTableRule result = new BindingTableRule();
         result.getTableRules().putAll(tableRules);
         return result;
     }
     
+    private void checkTheSameActualDataSourceNames(final Map<String, TableRule> tableRules, final String bindingTableGroup) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:205]: @cheese8 Can we choose the first tableRule as sample, and then judge whether other tableRules are same with sample?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 3]
```diff
@@ -192,11 +194,65 @@ private String getDefaultGenerateKeyColumn(final KeyGenerateStrategyConfiguratio
...
+        for (TableRule tableRule : tableRules.values()) {
+            ShardingStrategyConfiguration shardingStrategyConfiguration = databaseOrTable ? tableRule.getDatabaseShardingStrategyConfig() : tableRule.getTableShardingStrategyConfig();
+            if (null == shardingStrategyConfiguration) {
+                continue;
+            }
+            String shardingColumn = getShardingColumn(shardingStrategyConfiguration);
+            if (null == savedShardingColumn) {
+                savedShardingColumn = shardingColumn;
+            } else if (null != savedShardingColumn && !savedShardingColumn.equalsIgnoreCase(shardingColumn)) {
+                log.warn("All the route computations will only use the sharding strategy of the primary table, but the databaseShardingStrategyConfig on bindTable `%s` are inconsistent.",
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:233]: @cheese8 I think exception may be better.

**Review by strongduanmu — CHANGES_REQUESTED** [review]
@cheese8 Besides, can you add some unit test for this check logic?

**Review by strongduanmu — CHANGES_REQUESTED** [review]
Hi @cheese8, can you take a look at ci error?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 4]
```diff
@@ -203,6 +212,60 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
         result.getTableRules().putAll(tableRules);
         return result;
     }
+
+    private void checkTheSameActualDataSourceNamesAndActualTableIndex(final Collection<String> bindingTableGroups, final Collection<ShardingTableRuleConfiguration> tables) {
+        if (null == bindingTableGroups || bindingTableGroups.size() <= 0) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:217]: @cheese8 Do you think `bindingTableGroups.isEmpty()` is better?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 5]
```diff
@@ -203,6 +212,60 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+            return;
+        }
+        Iterator<String> itBindingTableGroups = bindingTableGroups.iterator();
+        Map<String, ShardingTableRuleConfiguration> shardingTableRuleConfigurationMap = tables.stream().collect(Collectors.toMap(ShardingTableRuleConfiguration::getLogicTable, Function.identity()));
+        while (itBindingTableGroups.hasNext()) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:222]: Can we use foreach here?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 6]
```diff
@@ -203,6 +212,60 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+                ShardingTableRuleConfiguration shardingTableRuleConfiguration = shardingTableRuleConfigurationMap.get(bindingTable);
+                if (null == shardingTableRuleConfiguration) {
+                    continue;
+                }
+                String actualDataNodes = shardingTableRuleConfiguration.getActualDataNodes();
+                if (null == actualDataNodes) {
+                    continue;
+                }
+                if (null == savedActualDataNodesList || savedActualDataNodesList.size() == 0) {
+                    savedActualDataNodesList = new InlineExpressionParser(actualDataNodes).splitAndEvaluateNotFlatten();
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:236]: @cheese8 Why not use bindingTableRules for judgement?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 7]
```diff
@@ -132,6 +140,7 @@ public ShardingRule(final AlgorithmProvidedShardingRuleConfiguration config, fin
         defaultKeyGenerateAlgorithm = null == config.getDefaultKeyGenerateStrategy()
                 ? RequiredSPIRegistry.getRegisteredService(KeyGenerateAlgorithm.class) : keyGenerators.get(config.getDefaultKeyGenerateStrategy().getKeyGeneratorName());
         defaultShardingColumn = config.getDefaultShardingColumn();
+        checkTheSameActualDataSourceNamesAndActualTableIndex(config.getBindingTableGroups(), config.getTables());
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:143]: Do you think `checkSameBindingTables` is better?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 1 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 3 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 2 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 4 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 8]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
         return result;
     }
     
+    private void checkSameBindingTables(final Collection<String> bindingTableGroups, final Map<String, BindingTableRule> bindingTableRules) {
+        if (null == bindingTableGroups || bindingTableGroups.isEmpty()) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:212]: @cheese8 Please remove this useless judge logic.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 9]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+            return;
+        }
+        for (String bindingTableGroup : bindingTableGroups) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:215]: Please rename bindingTableGroup with each.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 10]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+            List<String> bindingTableList = Splitter.on(",").trimResults().splitToList(bindingTableGroup.toLowerCase());
+            TableRule savedTableRule = null;
+            for (String bindingTable : bindingTableList) {
+                TableRule tableRule = bindingTableRules.get(bindingTable).getTableRules().get(bindingTable);
+                if (null == savedTableRule) {
+                    savedTableRule = tableRule;
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:221]: Do you think sampleTableRule is better?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 11]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+                } else {
+                    checkSameActualDatasourceNamesAndActualTableIndex(savedTableRule, tableRule, bindingTableGroup);
+                }
+            }
+        }
+    }
+    
+    private void checkSameActualDatasourceNamesAndActualTableIndex(final TableRule savedOne, final TableRule newOne, final String bindingTableGroup) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:229]: Please rename savedOne to sampleTableRule, rename newOne to tableRule.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 12]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+        if (!savedOne.getActualDatasourceNames().containsAll(newOne.getActualDatasourceNames())) {
+            throw new ShardingSphereConfigurationException("The actualDatasourceNames on bindingTableGroup `%s` are inconsistent", bindingTableGroup);
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:231]: Please split word or exception message.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 13]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+        }
+        checkSameAlgorithmOnDatabase(savedOne, newOne, savedOne.getActualDatasourceNames().stream().findFirst().get(), bindingTableGroup);
+        for (String dataSourceName : savedOne.getActualDatasourceNames()) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:234]: Please rename dataSourceName to each.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 14]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+            Collection<String> savedActualTableNames = savedOne.getActualTableNames(dataSourceName).stream().map(each -> substring(each)[1])
+                    .filter(each -> !each.isEmpty()).collect(Collectors.toList());
+            Collection<String> newOneActualTableNames = newOne.getActualTableNames(dataSourceName).stream().map(each -> substring(each)[1])
+                    .filter(each -> !each.isEmpty()).collect(Collectors.toList());
+            if (!savedActualTableNames.containsAll(newOneActualTableNames)) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:239]: @cheese8 Is this logic correct? The real table corresponding to the binding table should be different.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 15]
```diff
@@ -204,6 +208,105 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+                throw new ShardingSphereConfigurationException("The actualTableNames on bindingTableGroup `%s` are inconsistent", bindingTableGroup);
+            }
+            checkSameAlgorithmOnTable(savedOne, savedOne.getActualTableNames(dataSourceName).stream().findFirst().get(), newOne,
+                    newOne.getActualTableNames(dataSourceName).stream().findFirst().get(), bindingTableGroup);
+        }
+    }
+    
+    private void checkSameAlgorithmOnDatabase(final TableRule savedOne, final TableRule newOne, final String dataSourceName,
+                                              final String bindingTableGroup) {
+        Collection<String[]> args = new ArrayList<>();
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:249]: @cheese8 Please use more meaningful name for savedOne, newOne and args.

**Review by strongduanmu — CHANGES_REQUESTED** [review]
@cheese8 Thank you for your contribution, but this pr still has something needs to optimize. Besides, can we add some unit test to cover this logic change?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 9 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 10 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 11 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 12 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 13 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 15 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: ok

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 16]
```diff
@@ -216,6 +221,104 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+                    sampleTableRule = tableRule;
+                } else {
+                    checkSameActualDatasourceNamesAndActualTableIndex(sampleTableRule, tableRule, each);
+                }
+            }
+        }
+    }
+    
+    private void checkSameActualDatasourceNamesAndActualTableIndex(final TableRule sampleTableRule, final TableRule tableRule, final String bindingTableGroup) {
+        if (!CollectionUtils.isEqualCollection(sampleTableRule.getActualDatasourceNames(), tableRule.getActualDatasourceNames())) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:240]: @cheese8 Can we remove this util class? Replace it with `sampleTableRule.getActualDatasourceNames().equals( tableRule.getActualDatasourceNames()))`

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 17]
```diff
@@ -216,6 +221,104 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+                throw new ShardingSphereConfigurationException("The %s on bindingTableGroup `%s` are inconsistent", "actualTableNames", bindingTableGroup);
+            }
+            checkSameAlgorithmOnTable(sampleTableRule, sampleTableRule.getActualTableNames(each).stream().findFirst().get(), tableRule,
+                    tableRule.getActualTableNames(each).stream().findFirst().get(), bindingTableGroup);
+        }
+    }
+    
+    private void checkSameAlgorithmOnDatabase(final TableRule sampleTableRule, final TableRule tableRule, final String dataSourceName,
+                                              final String bindingTableGroup) {
+        Collection<String[]> algorithmExpressions = new ArrayList<>();
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:259]: @cheese8 LinkedList may be better.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 18]
```diff
@@ -216,6 +221,104 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+        String logicName = substring(dataSourceName)[0];
+        algorithmExpressions.add(new String[] {getAlgorithmExpression(sampleTableRule, true), logicName, getShardingColumn(sampleTableRule.getDatabaseShardingStrategyConfig())});
+        algorithmExpressions.add(new String[] {getAlgorithmExpression(tableRule, true), logicName, getShardingColumn(tableRule.getDatabaseShardingStrategyConfig())});
+        checkSameAlgorithmExpression(algorithmExpressions, "databaseShardingStrategyConfig", bindingTableGroup);
+    }
+    
+    private String getAlgorithmExpression(final TableRule tableRule, final boolean databaseOrTable) {
+        if (null == tableRule.getDatabaseShardingStrategyConfig() || null == tableRule.getTableShardingStrategyConfig()) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:267]: Is this logic wrong when check table sharding?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java [hunk 19]
```diff
@@ -216,6 +221,104 @@ private BindingTableRule createBindingTableRule(final String bindingTableGroup)
...
+            return "";
+        }
+        String shardingAlgorithmName = databaseOrTable ? tableRule.getDatabaseShardingStrategyConfig().getShardingAlgorithmName()
+                : tableRule.getTableShardingStrategyConfig().getShardingAlgorithmName();
+        ShardingAlgorithm shardingAlgorithm = shardingAlgorithms.get(shardingAlgorithmName);
+        return null == shardingAlgorithm ? "" : checkWithDefaultValue(shardingAlgorithm.getProps().getProperty("algorithm-expression"), "");
+    }
+    
+    private String checkWithDefaultValue(final String str, final String defaultValue) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java:276]: @cheese8 Can we remove this method with `StringUtils.defaultString()`?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 19 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: OK, it's better

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 18 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: Yes, logic fixed

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 17 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: OK

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java (hunk 16 above)
Comment (cheese8) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/ShardingRule.java]: OK

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**